hail/src/main/scala/is/hail/rvd/RVD.scala

```diff
@@ -534,21 +534,52 @@ class RVD(
     RVD(typ, newPartitioner, newRDD)
   }
 
-  def filter(p: (RegionValue) => Boolean): RVD =
-    RVD(typ, partitioner, crddBoundary.filter(p))
+  def filter(p: (RegionValue) => Boolean): RVD = {
+    filterWithContext((_, _) => (), (_: Any, rv) => p(rv))
+  }
+
+//  def filterWithContextNoFresh[C](makeContext: (Int, RVDContext) => C, f: (C, RegionValue) => Boolean): RVD = {
+//
+//    val rdd = this.crdd.rdd.mapPartitionsWithIndex{ (i, it) =>
+//      Iterator.single(
+//        (consumerCtx: RVDContext) => {
+//          val producerCtx = consumerCtx.freshContext
+//          val iteratorToFilter = it.flatMap(_(producerCtx))
+//          val c = makeContext(i, producerCtx)
+//          iteratorToFilter.filter { rv =>
+//            if (f(c, rv)) {
+//              producerCtx.region.move(consumerCtx.region)
+//              true
+//            }
+//            else {
+//              producerCtx.region.clear()
+//              false
+//            }
+//          }
+//        }
+//      )
+//    }
+//    val crdd = ContextRDD(rdd)
+//    RVD(this.typ, this.partitioner, crdd)
+//  }
 
   def filterWithContext[C](makeContext: (Int, RVDContext) => C, f: (C, RegionValue) => Boolean): RVD = {
-    mapPartitionsWithIndex(typ, { (i, context, it) =>
-      val c = makeContext(i, context)
-      it.filter { rv =>
-        if (f(c, rv))
-          true
-        else {
-          context.r.clear()
-          false
+
+    val crdd: ContextRDD[RVDContext, RegionValue] = this.crdd.usingFreshContext.cmapPartitionsWithIndex{ (i, consumerCtx: RVDContext, iteratorToFilter) => {
+          val c = makeContext(i, consumerCtx)
+          iteratorToFilter.filter { rv =>
```

Comment:
This looks close, but a bit messy.  Let's try to tighten it up

I don't think we should need a function like usingFreshContext.

There's always been a confusion between the context and the region value in my mind.  Thinking out load, let me try to clarify.  Context and region value both have regions.  What's the relationship between these regions?  The rule should be, if you call `it: (C) => Iterator[RV]` with `c`, then the region of region values returned by `it(c)` should be child regions of the region of `c`.

So, for example, in this code, I want to clear (or move) `c`, not `rv`, because there might be additional memory pointed to in `c` that's not protected by `rv`.

I think rv.region should never be used and new allocation should happen in c.region.  We could have made it Iterator[Long].

Also, I'm not sure we have quite the right interface to ContextRDD.  Here is how I would want to write it:

```
class ContextRDD[C, T]
  // I don't think this signature exists now but it is good.
  def cmapPartitions(f: (C, (C) => Iterator[T]) => Iterator[T]): ContextRDD[C, T]

class RVD
  val crdd: ContextRDD[C, RV]

  def filter(f: (C, RV) => Boolean): RVD =
    RVD(typ, partitioner,
      crdd.mapPartitions { case (consumerCtx, it) =>
	val c = consumerCtx.freshContext
	it(c).filter { rv =>
          // new allocation should go in c
	  val b = p(c, rv)
	  if (b)
	    consumerCtx.addChild(c)
	  c.clear()
	  b
	}
      })
```

Reply:
Thanks, I'll try this

Reply:
Also, agree that `rv.region` shouldn't be used, but I guess I am not sure if it is currently used, which is why in my `usingFreshContext` thing I move everything over to the region that will be on the context. Thought it might be helpful in the intermediate time before everything is converted.